PIPE2D-380: save wavelength fit data

PaulPrice opened the pull request:
A new class, WavelengthFitData, saves the wavelength fit data
(centroid in pixels, fit wavelength, actual wavelength) for
calculating statistics and for later QA.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
 
 class LineData(SimpleNamespace):
     """Data for a single reference line
-
```
There needs to be an empty line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
-        Measured wavelength of line in spectrum (nm).
-    actualWavelength : `float`
-        Actual wavelength of line (nm).
+    reflines : 'float'
```
Need backticks around the type.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
+        Pixel position calculated with detectormap.
+    fitPixelPos : `float`
+        Pixel position of reference line determined by the fit
+    fitWavelength : 'float'
```
Backticks.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
+        wavelength determined by the fit position.
+    fitPixelPosErr : `float`
+        Error of fit fitPixelPos 
+    fitWavelengthErr : 
```
Need type, in backticks.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
+        Error of fit fitPixelPos 
+    fitWavelengthErr : 
+        Error of fit Wavelength
+    wavelengthCorr : 'float'
```
Backticks.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
+        Error of fit Wavelength
+    wavelengthCorr : 'float'
+        Chevichev fit result 
+    status : int
```
Put type in backticks. Is it actually an int, or is the type `pfs.drp.stella.ReferenceLine.Status` (which can work like an int, but is actually a C++ enum)?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
-                                actualWavelength=actualWavelength)
 
+    def __init__(self, fiberId, reflines, nominalXPixelPos, fitXPixelPos, nominalPixelPos, fitPixelPos, fitWavelength, fitPixelPosErr, fitWavelengthErr, wavelengthCorr,status):
+        return super().__init__(fiberId=fiberId, reflines=reflines, nominalXPixelPos=nominalXPixelPos, fitXPixelPos=fitXPixelPos, nominalPixelPos=nominalPixelPos, fitPixelPos=fitPixelPos, fitWavelength=fitWavelength, fitPixelPosErr=fitPixelPosErr, fitWavelengthErr=fitWavelengthErr, wavelengthCorr=wavelengthCorr,status=status)
```
Line length limit is 110 chars.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
-    actualWavelength : `float`
-        Actual wavelength of line (nm).
+    reflines : 'float'
+        reference line wavelength (nm).         
```
This line and several below have trailing whitespace.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
 
 class WavelengthFitData:
     """Data characterising the quality of the wavelength fit for an image
-
```
Need a blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -51,19 +58,54 @@ def fiberId(self):
-    def pixels(self):
-        """Array of pixel positions (`numpy.ndarray` of `float`)"""
-        return np.array([ll.pixels for ll in self.lines])
+
```
Should be no blank line between `@property` and the method `def`.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -51,19 +58,54 @@ def fiberId(self):
+    def fitWavelength(self):
+        """Array of fit wavelengths in nm (`numpy.ndarray` of `float`)"""
+        return np.array([ll.fitWavelength for ll in self.lines])
+    @property
```
Need blank line between methods.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -51,19 +58,54 @@ def fiberId(self):
+
+    @property
+    def status(self):
+        """flags for status """
```
What's the return type?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -75,23 +117,23 @@ def __iter__(self):
 
     def residuals(self, fiberId=None):
         """Return wavelength residuals (nm)
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -75,23 +117,23 @@ def __iter__(self):
         ----------
         fiberId : `int`, optional
             Fiber identifier to select.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -75,23 +117,23 @@ def __iter__(self):
             Wavelength residuals (nm): measured - actual.
         """
-        return np.array([ll.measuredWavelength - ll.actualWavelength for ll in self.lines if
+
```
No blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -75,23 +117,23 @@ def __iter__(self):
+        return np.array([ll.fitWavelength - ll.reflines for ll in self.lines if
                          fiberId is None or ll.fiberId == fiberId])
 
+
```
Remove extra blank line (one and only one blank line between methods).

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -75,23 +117,23 @@ def __iter__(self):
     def mean(self, fiberId=None):
-        """Return the mean of wavelength residuals (nm)
 
+        """Return the mean of wavelength residuals (nm)
```
Remove blank line between `def` and docstring, and reinsert blank line between one-line summary and Parameters.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -106,47 +148,24 @@ def mean(self, fiberId=None):
 
     def stdev(self, fiberId=None):
         """Return the standard deviation of wavelength residuals (nm).
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -106,47 +148,24 @@ def mean(self, fiberId=None):
         ----------
         fiberId : `int`, optional
             Fiber identifier to select.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -15,26 +15,33 @@
+    def __init__(self, fiberId, reflines, nominalXPixelPos, fitXPixelPos, nominalPixelPos, fitPixelPos, fitWavelength, fitPixelPosErr, fitWavelengthErr, wavelengthCorr,status):
+        return super().__init__(fiberId=fiberId, reflines=reflines, nominalXPixelPos=nominalXPixelPos, fitXPixelPos=fitXPixelPos, nominalPixelPos=nominalPixelPos, fitPixelPos=fitPixelPos, fitWavelength=fitWavelength, fitPixelPosErr=fitPixelPosErr, fitWavelengthErr=fitWavelengthErr, wavelengthCorr=wavelengthCorr,status=status)
 
 class WavelengthFitData:
```
Need two blank lines between classes.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -106,47 +148,24 @@ def mean(self, fiberId=None):
         return self.residuals(fiberId).std()
 
-    @classmethod
-    def fromSpectrumSet(cls, spectrumSet, detectorMap):
```
Why get rid of this method? It seems like a good, self-contained way of constructing the object.

nakakeigo replied:
I modified the code so that I can get the info directly from calibrateWavelengthTask (or I should duplicate most of the line) and this method is not used to get the info.
I’ll fix it if needed.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -106,47 +148,24 @@ def mean(self, fiberId=None):
     @classmethod
     def readFits(cls, filename):
         """Read from file
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -106,47 +148,24 @@ def mean(self, fiberId=None):
         ----------
         filename : `str`
             Name of file from which to read.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -155,24 +174,39 @@ def readFits(cls, filename):
+            fitPixelPosErr = hdu.data["fitPixelPosErr"]
+            fitWavelengthErr = hdu.data["fitWavelengthErr"]        
+            wavelengthCorr = hdu.data["wavelengthCorr"]
+            status =hdu.data["status"] 
```
Need space either side of `=`.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -155,24 +174,39 @@ def readFits(cls, filename):
+            wavelengthCorr = hdu.data["wavelengthCorr"]
+            status =hdu.data["status"] 
+        
+        return cls([LineData(*args) for args in zip(fiberId, reflines, nominalXPixelPos, fitXPixelPos, nominalPixelPos, fitPixelPos, fitWavelength, fitPixelPosErr, fitWavelengthErr, wavelengthCorr, status)])
```
Exceeds line length limit.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -155,24 +174,39 @@ def readFits(cls, filename):
 
     def writeFits(self, filename):
         """Write to file
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -155,24 +174,39 @@ def readFits(cls, filename):
-            astropy.io.fits.Column(name="actualWavelength", format="D", array=self.actualWavelength),
+            astropy.io.fits.Column(name="reflines", format="D", array=self.reflines),
+            astropy.io.fits.Column(name="nominalXPixelPos", format="D", array=self.nominalXPixelPos),
+            astropy.io.fits.Column(name="fitXPixelPos", format="D", array=self.fitXPixelPos),  
```
Do pixel positions need double precision?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -202,23 +236,24 @@ def __init__(self, *args, **kwargs):
-        """Fit wavelength solution for a spectrum
-
+    def fitWavelengthSolution(self, spec, detectorMap,rng=np.random):
+    
```
No blank line here.
Need space after comma in line above.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -202,23 +236,24 @@ def __init__(self, *args, **kwargs):
-        spec : `pfs.drp.stella.Spectrum`
-            Spectrum to fit; updated with solution.
+        spectrumSet : `pfs.drp.stella.SpectrumSet`
+            SpectrumSet to fit; updated with solution.
```
In general, you should avoid making changes like this that aren't necessary. If you think they are necessary, they should go in a separate commit.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -202,23 +236,24 @@ def __init__(self, *args, **kwargs):
             Mapping of wl,fiber to detector position; updated with solution.
         rng : `numpy.random.RandomState`
             Random number generator, for reserving some lines from the fit.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -202,23 +236,24 @@ def __init__(self, *args, **kwargs):
         -------
-        wavelengthCorr : `np.polynomial.chebyshev.Chebyshev`
-            Wavelength solution.
+        : wlFitData
```
Fix the return list so it's accurate.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -202,23 +236,24 @@ def __init__(self, *args, **kwargs):
-            Wavelength solution.
+        : wlFitData
         """
+
```
No blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -234,6 +269,7 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         #
         wavelength = np.array([rl.wavelength for rl in refLines])
         status = np.array([rl.status for rl in refLines])
+        flag = np.array([rl.status for rl in refLines])
```
Why change the name of this variable? It seems unnecessary.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -249,46 +285,50 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
                 #
+
                 goodIndices = np.where(good)[0]
+
```
Don't introduce these blank lines unnecessarily.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -249,46 +285,50 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+
+            if good.sum() == 0:
+                self.log.warn("All points were clipped for fiberId %d; disabled clipping" % fiberId)
+                good[:] = True
```
With this change you have removed the iterative fitting. It needs to be undone.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -78,46 +285,50 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+            flag[i] = good[i]
             if not good[i]:
                 rl.status |= rl.Status.CLIPPED
+                flag[i]=3
```
And need space either side of `=`.

nakakeigo replied:
removed flag and used rl.Status value.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -78,46 +285,50 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+                flag[i]=3
             if reserved[i]:
                 rl.status |= rl.Status.RESERVED
+                flag[i]=5
```
And need space either side of `=`.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -249,46 +285,50 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         #
         # Correct the initial wavelength solution
         #
+
```
Unnecessary addition.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -302,11 +342,11 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         #
         if self.config.resetSlitDy:
-            offsets = detectorMap.getSlitOffsets(fiberId)
+            offsets = detectorMap.getSlitOffsets(wlData.fiberId)
```
Why is this change necessary?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -302,11 +342,11 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
             offsets[detectorMap.FIBER_DY] = 0
             detectorMap.setSlitOffsets(fiberId, offsets)
-
+            
```
Remove added whitespace.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
                 dy = -int(-dy)
                 spec.wavelength[dy:] = spec.wavelength[:-dy]
-
+                
```
Remove added whitespace.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+        detectorMap.setWavelength(fiberId, rows, spec.wavelength)
         diff = detectorMap.getWavelength(fiberId) - spec.wavelength
         self.log.info("Fiber %d: wavelength correction %f +/- %f nm" % (fiberId, diff.mean(), diff.std()))
-        detectorMap.setWavelength(fiberId, rows, spec.wavelength)
```
Moving this line makes the above log line useless, because now `detectorMap.getWavelength(fiberId)` equals `spec.wavelength`.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+            fitXPosition = detectorMap.findPoint(fiberId, fitWavelength[i])[0]
+            nominalXPosition = detectorMap.findPoint(fiberId, wavelength[i])[0]
+            line=LineData(fiberId, wavelength[i], nominalXPosition, fitXPosition, nominalPixelPos[i], rl.fitPosition, fitWavelength[i], rl.fitPositionErr, fitWavelengthErr[i], wavelengthCorr(nominalPixelPos[i]),flag[i])
+            lines.append(line)
```
This block should go in a `WavelengthFitData` constructor.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+            nominalXPosition = detectorMap.findPoint(fiberId, wavelength[i])[0]
+            line=LineData(fiberId, wavelength[i], nominalXPosition, fitXPosition, nominalPixelPos[i], rl.fitPosition, fitWavelength[i], rl.fitPositionErr, fitWavelengthErr[i], wavelengthCorr(nominalPixelPos[i]),flag[i])
+            lines.append(line)
+        return [lines, wavelengthCorr]
```
If you have to return more than one thing from a method, please use `lsst.pipe.base.Struct`, so the elements are named.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
 
-        return wavelengthCorr
 
     def plot(self, spec, detectorMap, wavelengthCorr):
```
Only one line between methods.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -315,16 +355,24 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
 
     def plot(self, spec, detectorMap, wavelengthCorr):
         """Plot fit results
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -425,7 +473,6 @@ def plot(self, spec, detectorMap, wavelengthCorr):
 
     def measureStatistics(self, spectrumSet, detectorMap):
         """Measure some statistics about the solution
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -445,9 +492,7 @@ def measureStatistics(self, spectrumSet, detectorMap):
 
     def run(self, spectrumSet, detectorMap, seed=1):
         """Run the wavelength calibration
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -445,9 +492,7 @@ def measureStatistics(self, spectrumSet, detectorMap):
         """Run the wavelength calibration
-
         Assumes that line identification has been done already.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -456,7 +501,6 @@ def run(self, spectrumSet, detectorMap, seed=1):
             Mapping of wl,fiber to detector position.
         seed : `int`
             Seed for random number generator.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
+
         solutions = []
+        
         for spec in spectrumSet:
```
Too many blank lines above that serve no purpose.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
-            wavelengthCorr = self.fitWavelengthSolution(spec, detectorMap, rng)
+            results = self.fitWavelengthSolution(spec, detectorMap, rng)
+            wavelengthCorr =results[1]
+            lines =results[0]
```
Need space either side of `=` (on this and the previous line).

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
+        
         return pipeBase.Struct(solutions=solutions, wlFitData=wlFitData)
+        
+    def robustStdev(array):
```
Lots of trailing whitespace on above lines.

PaulPrice replied:
This is supposed to be a free function, not a method (note there's no `self` parameter). Looks like I put it in the wrong place, but it doesn't belong here. Put it at the end of the file.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
             display.erase()
 
+
+        lineDatas = []
```
`lineDatas` isn't good English. `lineData` would be better.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
 
     def runDataRef(self, dataRef, spectrumSet, detectorMap, seed=1):
         """Run the wavelength calibration
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,41 +513,46 @@ def run(self, spectrumSet, detectorMap, seed=1):
         """Run the wavelength calibration
-
         Assumes that line identification has been done already.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -514,14 +563,14 @@ def runDataRef(self, dataRef, spectrumSet, detectorMap, seed=1):
             Mapping of wl,fiber to detector position.
         seed : `int`
             Seed for random number generator.
-
```
Need blank line here.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -514,14 +563,14 @@ def runDataRef(self, dataRef, spectrumSet, detectorMap, seed=1):
         wlFitData : `WavelengthFitData`
             Data on quality of the wavelength fit.
         """
+
```
No blank line here.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -18,37 +18,41 @@ class WavelengthFitDataTestCase(lsst.utils.tests.TestCase):
+        self.wavelengthCorr = (np.random.uniform(size=self.fiberId.shape)*scale + self.wlMin)*0.01
+        self.reflines = np.random.uniform(size=self.fiberId.shape)*scale + self.wlMin
+        self.status = np.random.uniform(size=self.fiberId.shape)
+        self.lines = [LineData(*args) for args in zip(self.fiberId, self.reflines,self.nominalXPixelPos, self.fitXPixelPos, self.nominalPixelPos, self.fitPixelPos, self.fitWavelength, self.fitPixelPosErr, self.fitWavelengthErr,self.wavelengthCorr, self.status)]
```
Exceeds line length limit.
Need space after comma.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -18,37 +18,41 @@ class WavelengthFitDataTestCase(lsst.utils.tests.TestCase):
 
     def assertWavelengthFitData(self, wlFitData, atol=0.0):
         """Check that the WavelengthFitData is what we expect
-
```
Need blank line here.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -18,37 +18,41 @@ class WavelengthFitDataTestCase(lsst.utils.tests.TestCase):
         self.actualWavelength = np.random.uniform(size=self.fiberId.shape)*scale + self.wlMin
-        self.lines = [LineData(*args) for args in zip(self.fiberId, self.pixels, self.measuredWavelength,
-                                                      self.actualWavelength)]
+        self.fitWavelengthErr = (np.random.uniform(size=self.fiberId.shape)*scale + self.wlMin)*0.0
```
Multiply by zero? Why not use `np.zeros`?

nakakeigo replied:
It seems 1 (0.01) is mistakenly removed (1% Error is assigned) .

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -18,37 +18,41 @@ class WavelengthFitDataTestCase(lsst.utils.tests.TestCase):
             Object to check.
         atol : `float`
-            Absolute tolerance for ``measuredWavelength`` (might not be exact
-            if calculated).
```
This line got removed.

PaulPrice replied:
It's still removed.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -18,37 +18,41 @@ class WavelengthFitDataTestCase(lsst.utils.tests.TestCase):
-        self.assertFloatsAlmostEqual(wlFitData.measuredWavelength, self.measuredWavelength, atol=atol)
-        self.assertFloatsEqual(wlFitData.actualWavelength, self.actualWavelength)
-
+        self.assertFloatsAlmostEqual(wlFitData.fitWavelength, self.fitWavelength, atol=atol)
```
Why did you remove three tests and only add one? You added more functionality, so you should be adding more tests.

nakakeigo replied:
I added several lines. 
I'll make more lines if needed.

PaulPrice replied:
Please add tests for all the elements of `WavelengthFitData`.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -67,25 +71,28 @@ def testFromSpectrumSet(self):
+            wavelengthValues.append(np.array([self.wlMin - 1, self.wlMin, self.fitWavelength[ii],
                                               self.wlMax, self.wlMax + 1],
                                              dtype=np.float32))
+            linedata=LineData(ss.fiberId, line, 0, 0, line.fitPosition, 0, line.wavelength,0, 0, 0,line.status)
```
Need space either side of `=` and after commas.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -109,7 +116,9 @@ def setup_module(module):
     setup_module(sys.modules["__main__"])
+
     from argparse import ArgumentParser
+
```
Don't add these extra spaces.

RobertLuptonTheGood commented:
Why are we scaling by 0.01?  If we want to put out diagnostics in % that's fine, but let's not make assumptions like that for the numbers we calculate

PaulPrice commented:
It's a test, and 1% error is just what has been chosen to set up the test. I don't believe the choice of value is all that important for this particular test either.

RobertLuptonTheGood commented:
I'd still rather put the % in the test value, not the self.fitWavelengthErr

PaulPrice commented:
Yes, it should be a parameter.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -216,13 +315,15 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         fiberId = spec.getFiberId()
         lam = detectorMap.getWavelength(fiberId)
-        nmPerPix = (lam[-1] - lam[0])/(rows[-1] - rows[0])
+        nmPerPix = (lam[-1] - lam[0]) / (rows[-1] - rows[0])
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -235,9 +336,9 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         status = np.array([rl.status for rl in refLines])
         fitPosition = np.array([rl.fitPosition for rl in refLines])
-        nominalPixelPos = np.array([(rl.wavelength - lam[0])/nmPerPix for rl in refLines])
+        nominalPixelPos = np.array([(rl.wavelength - lam[0]) / nmPerPix for rl in refLines])
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -235,9 +336,9 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+        nominalPixelPos = np.array([(rl.wavelength - lam[0]) / nmPerPix for rl in refLines])
         fitWavelength = np.array([detectorMap.findWavelength(fiberId, rl.fitPosition) for rl in refLines])
-        fitWavelengthErr = np.array([rl.fitPositionErr*nmPerPix for rl in refLines])
+        fitWavelengthErr = np.array([rl.fitPositionErr * nmPerPix for rl in refLines])
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
             y = wavelength - fitWavelength
-
-            yerr = np.hypot(fitWavelengthErr, self.config.pixelPosErrorFloor*nmPerPix)
+            yerr = np.hypot(fitWavelengthErr, self.config.pixelPosErrorFloor * nmPerPix)
```
Unnecessary changes.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
-
             wavelengthCorr = np.polynomial.chebyshev.Chebyshev.fit(
-                x[use], y[use], self.config.order, domain=[0, len(spec.wavelength) - 1], w=1/yerr[use])
+                x[use], y[use], self.config.order, domain=[0, len(spec.wavelength) - 1], w=1 / yerr[use])
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
                 resid = y - yfit
                 stdev = robustStdev(resid)
-                good &= (np.fabs(resid) < nSigma*np.where(yerr > stdev, yerr, stdev))
+                good &= (np.fabs(resid) < nSigma * np.where(yerr > stdev, yerr, stdev))
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
         spec.wavelength = wavelengthCorr(rows).astype('float32') + detectorMap.getWavelength(fiberId)
 
-        rmsFit = np.sqrt(np.sum(((y - yfit)**2)[use])/(use.sum() - self.config.order))
+        rmsFit = np.sqrt(np.sum(((y - yfit)**2)[use]) / (use.sum() - self.config.order))
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
-                      fiberId, rmsFit, rmsFit/nmPerPix, good.sum(), len(refLines),
-                      rmsReserved, rmsReserved/nmPerPix, reserved.sum(),
+                      fiberId, rmsFit, rmsFit / nmPerPix, good.sum(), len(refLines),
+                      rmsReserved, rmsReserved / nmPerPix, reserved.sum(),
```
Unnecessary changes.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -262,41 +363,39 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
+                    if not good[i]:
+                        rl.status |= rl.Status.CLIPPED
+                    if reserved[i]:
+                        rl.status |= rl.Status.RESERVED
```
Why has this changed? It's not necessary to do this every iteration, right?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -316,15 +415,15 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
                 spec.wavelength[dy:] = spec.wavelength[:-dy]
 
-        diff = detectorMap.getWavelength(fiberId) - spec.wavelength
-        self.log.info("Fiber %d: wavelength correction %f +/- %f nm" % (fiberId, diff.mean(), diff.std()))
```
Why remove this line? Isn't it useful?

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -216,13 +315,15 @@ def fitWavelengthSolution(self, spec, detectorMap, rng=np.random):
-            Wavelength solution.
+        wlFitData : list of `LineData`
+            Data for wavelength solution evaluation
+        wavelengthCorr : `float`
```
I don't think this change is accurate.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,36 +568,24 @@ def run(self, spectrumSet, detectorMap, seed=1):
+        lineData = []
+
         solutions = []
+
```
Unnecessary change.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -469,36 +568,24 @@ def run(self, spectrumSet, detectorMap, seed=1):
-    """
-    lq, uq = np.percentile(array, (25.0, 75.0))
-    return 0.741*(uq - lq)
-
```
This change should be a separate commit.

PaulPrice commented on python/pfs/drp/stella/calibrateWavelengthsTask.py:
```diff
@@ -522,7 +609,26 @@ def runDataRef(self, dataRef, spectrumSet, detectorMap, seed=1):
         wlFitData : `WavelengthFitData`
             Data on quality of the wavelength fit.
         """
+
```
Unnecessary change.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -67,25 +79,38 @@ def testFromSpectrumSet(self):
                                    lsst.afw.geom.Extent2I(self.length, self.length))
-        detMap = DetectorMap(bbox, self.fiberId, centerKnots, centerValues, wavelengthKnots, wavelengthValues)
+        detMap = DetectorMap(bbox, self.fiberId,
+                             centerKnots, centerValues, wavelengthKnots, wavelengthValues)
```
Unnecessary change.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -67,25 +79,38 @@ def testFromSpectrumSet(self):
+
+            linedata = LineData(ss.fiberId, line.wavelength, 0, 0, 0,
+                                line.fitPosition, detMap.getWavelength(ss.fiberId, line.fitPosition),
+                                0, 0, 0, line.status)
```
Please use actual values for the ones you've set to `0` here, and then test that they are set properly.

PaulPrice commented on tests/test_WavelengthFitData.py:
```diff
@@ -109,6 +134,7 @@ def setup_module(module):
 
 if __name__ == "__main__":
     setup_module(sys.modules["__main__"])
+
```
Unnecessary change.

PaulPrice commented:
@nakakeigo: I cleaned up your additions. I changed the spelling on some of the elements (to reduce the potential for confusion). Please have a look and make sure you're happy with the result, and then we'll merge.

nakakeigo commented:
Thank you very much for cleaning up my dirty addition!. I'm happy with the update.
